### PR TITLE
[SPARK-37083][PYTHON] Inline type hints for python/pyspark/accumulators.py

### PR DESCRIPTION
**Pull request by dchvn**
### What changes were proposed in this pull request?
Inline type hints for python/pyspark/accumulators.py

### Why are the changes needed?
We can take advantage of static type checking within the functions by inlining the type hints.

### Does this PR introduce _any_ user-facing change?
No

### How was this patch tested?
Existing tests

### CODE REVIEW

File: python/pyspark/_typing.pyi [hunk 1]
```diff
@@ -21,11 +21,14 @@ from typing_extensions import Protocol
 
 F = TypeVar("F", bound=Callable)
 T = TypeVar("T", covariant=True)
+U = TypeVar("U", bound="SupportsIAdd")
```
Comment (zero323) [inline python/pyspark/_typing.pyi:24]: Why would add bound for protocol?

File: python/pyspark/_typing.pyi [hunk 2]
```diff
@@ -21,11 +21,14 @@ from typing_extensions import Protocol
...
 
 PrimitiveType = Union[bool, float, int, str]
 
 class SupportsIAdd(Protocol):
-    def __iadd__(self, other: SupportsIAdd) -> SupportsIAdd: ...
+    def __add__(self: U, other: U) -> U: ...
+
+    def __iadd__(self: U, other: U) -> U: ...
```
Comment (zero323) [inline python/pyspark/_typing.pyi:31]: Also, why this change of signature?

File: python/pyspark/_typing.pyi [hunk 3]
```diff
@@ -21,11 +21,14 @@ from typing_extensions import Protocol
 
 F = TypeVar("F", bound=Callable)
 T = TypeVar("T", covariant=True)
+U = TypeVar("U", bound="SupportsIAdd")
 
 PrimitiveType = Union[bool, float, int, str]
 
 class SupportsIAdd(Protocol):
-    def __iadd__(self, other: SupportsIAdd) -> SupportsIAdd: ...
+    def __add__(self: U, other: U) -> U: ...
```
Comment (xinrong-databricks) [inline python/pyspark/_typing.pyi:29]: Why do we need `__add__` in `SupportsIAdd`?

File: python/pyspark/_typing.pyi (hunk 1 above)
Comment (dchvn) [reply python/pyspark/_typing.pyi:24]: Should we not add bound `SupportsIAdd` for `U` which can make a constraint for types `U` can be ?

File: python/pyspark/accumulators.py [hunk 4]
```diff
@@ -176,44 +193,44 @@ class AccumulatorParam(object):
...
+    def __init__(self, zero_value: U):
         self.zero_value = zero_value
 
-    def zero(self, value):
+    def zero(self, value: U) -> U:
         return self.zero_value
 
-    def addInPlace(self, value1, value2):
+    def addInPlace(self, value1: U, value2: U) -> U:
         value1 += value2
```
[the remark this answers is not in the document]
Comment (zero323) [reply python/pyspark/accumulators.py:226]: `__add__` is not necessary for `__iadd__`. You can implement the latter one without the former. However, Python data model allows fallbacks from `x += y` to `x = x + y` when only `__add__` is implemented).

So if addition of `__add__` makes type checker happy, it might indicate a deeper problem, probably with the protocol

File: python/pyspark/_typing.pyi [hunk 5]
```diff
@@ -21,11 +21,14 @@ from typing_extensions import Protocol
 
 F = TypeVar("F", bound=Callable)
 T_co = TypeVar("T_co", covariant=True)
+T = TypeVar("T", covariant=True)
```
Comment (zero323) [inline python/pyspark/_typing.pyi:24]: Nit, this shouldn't be here.

File: python/pyspark/accumulators.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (zero323) [reply python/pyspark/accumulators.py:226]: OK, so reducing this a bit:

```python
# main.py
 
from __future__ import annotations
from typing import Generic, Protocol, TypeVar     
 
 
T = TypeVar("T")
U = TypeVar("U", bound=SupportsIAdd)

 
class SupportsIAdd(Protocol):
    def __iadd__(self: T, other: T) -> T: ...        

 
class AddingAccumulatorParam(Generic[U]):
    def addInPlace(self, value1: U, value2: U) -> U:
        value1 += value2
        return value1

```

```
main.py:17: error: Unsupported left operand type for + ("U")
Found 1 error in 1 file (checked 1 source file)
```

but after replacing `__iadd__` with `__add__` passes 

```python
# main.py
 
from __future__ import annotations
from typing import Generic, Protocol, TypeVar     
 
 
T = TypeVar("T")
U = TypeVar("U", bound=SupportsIAdd)

 
class SupportsIAdd(Protocol):
    def __add__(self: T, other: T) -> T: ...         

 
class AddingAccumulatorParam(Generic[U]):
    def addInPlace(self, value1: U, value2: U) -> U:
        value1 += value2
        return value1
```

I honestly don't see why, but there are all kinds of subtleties around `__iadd__`, `__radd__` and `__add__`. Anyone?

[the remark this answers is not in the document]
Comment (zero323) [reply python/pyspark/accumulators.py:226]: It seems that the problem is somehow related to combination of `__iadd__` and `TypeVar`. Further reducing the example ‒ this still fails 

```python
# main.py

from __future__ import annotations
from typing import Any, TypeVar              


U = TypeVar("U", bound="SupportsIAdd")        


class SupportsIAdd:          
    def __iadd__(self: U, other: Any) -> U: ...                 


class AddingAccumulatorParam:            
    def addInPlace(self, value1: U, value2: U) -> U: 
        value1 += value2
        return value1
````

but substituting `U` with `SupportsIAdd`:


```python
# main.py

from __future__ import annotations
from typing import Any, TypeVar              


U = TypeVar("U", bound="SupportsIAdd")        


class SupportsIAdd:          
    def __iadd__(self, other: Any) -> SupportsIAdd: ...         


class AddingAccumulatorParam:            
    def addInPlace(self, value1: SupportsIAdd, value2: SupportsIAdd) -> SupportsIAdd:            
        value1 += value2
        return value1

```

makes mypy happy. It don't think it helps us to resolve it here, but narrows down the scope, in  case we want to escalate this to mypy folks.

**Comment by zero323** [comment]
Unrelated to this specific PR ‒ we all use force pushes from time to time, especially when linear history is preferred (and we `rebase` to the `HEAD` of the main branch),  but if you regulary keep changing the history during the review, it  makes it hard to follow, especially when you squash the commits on the way. 

So please, use it sparingly and limit the intervention to what's really necessary, once you leave `[WIP]`  / `draft` stage.

File: python/pyspark/_typing.pyi (hunk 5 above)
Comment (itholic) [reply python/pyspark/_typing.pyi:24]: qq: Why do we have both `T_co` and `T` for the same type ??

Comment (dchvn) [reply python/pyspark/_typing.pyi]: updated. thanks!

**Comment by dchvn** [comment]
@zero323 Do you have any ideas to resolve this PR? Should I keep both `__add__` and `__iadd__` or follow your suggestion?

**Comment by zero323** [comment]
> @zero323 Do you have any ideas to resolve this PR? 

Not at the moment and I don't have much time to investigate it further. If you no one has any inputs, and you feel like it, you can try to reduce the [snippets](https://github.com/apache/spark/pull/34363#discussion_r760622451) and ask mypy folks, if that's expected behavior for some reason, or just  a bug.

If the latter, we can just `type: ingore`, if the former we can try to think about other approaches.

> Should I keep both `__add__` and `__iadd__` 

That's clearly not what we expect here. There is no requirement for us to have both `__iadd__` and `__add__` implemented. We could be more subtle at some point and have something like `Union[SupportsIAdd, SupportsAdd]`, but that's really not the point here.

**Comment by zero323** [comment]
> you can try to reduce the [snippets](https://github.com/apache/spark/pull/34363#discussion_r760622451) and ask mypy folks,

If you go this way, please link the discussion here, so we're all aware of it.

File: python/pyspark/accumulators.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (zero323) [reply python/pyspark/accumulators.py]: Posted on mypy gitter: https://gitter.im/python/typing?at=61b87c2f197fa95a1cc2c2a4

[the remark this answers is not in the document]
Comment (zero323) [reply python/pyspark/accumulators.py]: And corresponding mypy issue https://github.com/python/mypy/issues/11745

**Comment by HyukjinKwon** [comment]
sorry I completely missed the context here. What's the status now?

**Comment by zero323** [comment]
> sorry I completely missed the context here. What's the status now?

So far, there is no response to issue on mypy tracker, but we can be fairly sure that the issue that blocks this one is actually a bug. As far as I am concerned, we should resolve the conflicts, add `type: ignore` on `+=`, double check things and we're good for merge.

**Comment by dchvn** [comment]
> > sorry I completely missed the context here. What's the status now?
> 
> So far, there is no response to issue on mypy tracker, but we can be fairly sure that the issue that blocks this one is actually a bug. As far as I am concerned, we should resolve the conflicts, add `type: ignore` on `+=`, double check things and we're good for merge.

used  `# type: ignore[operator]`

**Review by zero323 — APPROVED** [review]

**Comment by zero323** [comment]
Merged into master.

**Comment by HyukjinKwon** [comment]
Thanks all!
